src/libs/sfdk/dockervirtualmachine.cpp

```diff
@@ -0,0 +1,531 @@
+    VirtualMachineInfo info = {};
+
+    if (document.isNull())
+        return info;
+
+    QJsonObject labels = document.object();
+
+    auto parsePort = [=](VirtualMachinePrivate::ReservedPort which, quint16& port) {
+        QString key = label(which);
+        if (labels.contains(key))
+            port = labels.value(key).toString().toUShort();
+    };
+
+    parsePort(VirtualMachinePrivate::SshPort, info.sshPort);
+    parsePort(VirtualMachinePrivate::WwwPort, info.wwwPort);
+
+    auto parsePath = [=](VirtualMachinePrivate::SharedPath which, QString& path) {
+        QString key = label(which);
+        if (labels.contains(key))
+            path = labels.value(key).toString();
+    };
+
+    parsePath(VirtualMachinePrivate::SharedSrc, info.sharedSrc);
+    parsePath(VirtualMachinePrivate::SharedSsh, info.sharedSsh);
+    parsePath(VirtualMachinePrivate::SharedHome, info.sharedHome);
+    parsePath(VirtualMachinePrivate::SharedConfig, info.sharedConfig);
+    parsePath(VirtualMachinePrivate::SharedTargets, info.sharedTargets);
+
+    info.memorySizeMb = VirtualMachine::availableMemorySizeMb();
+    info.cpuCount = VirtualMachine::availableCpuCount();
+
+    return info;
+}
+
+QString DockerVirtualMachinePrivate::label(VirtualMachinePrivate::SharedPath which)
+{
+    QString label;
+    switch (which) {
+    case SharedSrc:
+        label = Constants::BUILD_ENGINE_SHARED_SRC;
+        break;
+    case SharedHome:
+        label = Constants::BUILD_ENGINE_SHARED_HOME;
+        break;
+    case SharedConfig:
+        label = Constants::BUILD_ENGINE_SHARED_CONFIG;
+        break;
+    case SharedSsh:
+        label = Constants::BUILD_ENGINE_SHARED_SSH;
+        break;
+    case SharedTargets:
+        label = Constants::BUILD_ENGINE_SHARED_TARGET;
+    break;
+    }
+    return label;
+}
+
+QString DockerVirtualMachinePrivate::label(VirtualMachinePrivate::ReservedPort which)
+{
+    QString label;
+    switch (which) {
+    case SshPort:
+        label = Constants::BUILD_ENGINE_SSH_PORT;
+        break;
+    case WwwPort:
+        label = Constants::BUILD_ENGINE_WWW_PORT;
+        break;
+    }
+    return label;
+}
+
+
+Q_GLOBAL_STATIC(QString, s_dockerPath)
+
+QString DockerVirtualMachinePrivate::dockerPath()
+{
+    if (s_dockerPath()->isNull()) {
+        QStringList searchPaths = QProcessEnvironment::systemEnvironment()
+            .value("PATH").split(QChar(':'));
+        foreach (const QString &searchPath, searchPaths) {
+            QDir dir(searchPath);
+            if (dir.exists(DOCKER)) {
+                s_dockerPath()->operator=(dir.absoluteFilePath(DOCKER));
+                break;
+            }
+        }
+    }
+    return *s_dockerPath();
+}
+
+QStringList DockerVirtualMachinePrivate::listedImages(const QString &output)
+{
+    QStringList images;
+    QStringList lines = output.split(QRegularExpression("[\r\n]"), QString::SkipEmptyParts);
+    for (QString line : lines) {
```

Comment:
const QStringList lines, const QString &line

Reply:
fixed

Reply:
Oops, misplaced `&` yet.

Reply:
Fixed this and a few other occurrences as well